solcore/absorption_calculator/tmm_core_vec.py

```diff
@@ -961,16 +1027,72 @@ def inc_find_absorp_analytic_fn(layer, inc_data):
     inc_data is output of incoherent_main()
     """
     j = inc_data['stack_from_all'][layer]
-    if isnan(j):
+    if any(isnan(j)):
         raise ValueError('layer must be coherent for this function!')
     [stackindex, withinstackindex] = j
     forwardfunc = absorp_analytic_fn()
     forwardfunc.fill_in(inc_data['coh_tmm_data_list'][stackindex],
                         withinstackindex)
-    forwardfunc.scale(inc_data['stackFB_list'][stackindex][0])
+    forwardfunc.scale(inc_data['stackFB_list'][:, stackindex, 0])
     backfunc = absorp_analytic_fn()
     backfunc.fill_in(inc_data['coh_tmm_bdata_list'][stackindex],
                      -1 - withinstackindex)
-    backfunc.scale(inc_data['stackFB_list'][stackindex][1])
+    #print('before scale', backfunc.A2, backfunc.A1)
+    backfunc.scale(inc_data['stackFB_list'][:, stackindex, 1])
+    #print('after scale', backfunc.A2, backfunc.A1)
     backfunc.flip()
+    #print('after flip', backfunc.A2, backfunc.A1)
     return forwardfunc.add(backfunc)
+
+
+def inc_position_resolved(layer, dist, inc_tmm_data, coherency_list, widths, alphas):
+    """
+    This function is vectorized. Analogous to position_resolved, but
+    for layers (incoherent or coherent) in (partly) incoherent stacks.
+    This is a new function, not from Steven Byrnes' tmm package.
+    Starting with output of inc_tmm(), calculate the Poynting vector
+    and absorbed energy density a distance "dist" into layer number "layer"
+    """
+
+    layers = list(set(layer)) # unique layer indices
+    A_per_layer = np.array(inc_absorp_in_each_layer(inc_tmm_data))
+    fraction_reaching = 1 - np.cumsum(A_per_layer, axis = 0)
+    A_local = np.zeros((len(alphas[0]), len(dist)))
+    for i, l in enumerate(layers):
+        if coherency_list[l] == 'c':
+            #print(l, 'c')
+            fn = inc_find_absorp_analytic_fn(l, inc_tmm_data)
+            A_local[:, layer == l] = fn.run(dist[layer == l])
+
+            #print('c', A_local[:, layer == l])
+        else:
+            #print(l, 'i')
+            A_local[:, layer == l] = beer_lambert(widths[l]*1e-9, alphas[l]*1e9, fraction_reaching[i], dist[layer == l]*1e-9)
+            #print('i', A_local[:, layer == l])
+
+    #A_local[np.isnan(A_local)] = 0
+    return A_local
+
+
+def beer_lambert(width, alphas, fraction, dist):
```

Comment:
This function probably needs a proper docstring and have most of the commented print statements and inline comments removed.

Reply:
Added (also realised as a result of looking at the function that it, and inc_position_resolved, had an unnecessary argument, which I removed)